NodeInfo remove main identity

kasiastreich opened the pull request:
Preparation for getting rid of services + supporting multiple identities on the node.
I think this work will be influenced by Ross anonymised identity. For now signing identity was moved into `ServiceHub.legalIdentity`
* I removed separate SERVICES_PREFIX from artemis queues, because it wasn't necessary (when deploying bridges we query `NetworkMapCache` to see how many nodes we have associated with this key).
* Problem remains with SSL connections, design doc for new network map was clear about removing `SingleMessageRecipient` from `NodeInfo` and changing it to `NetworkHostAndPort`, but I am not convinced to that, maybe we should have IP address -> legalIdentity, then check it on connection.
* Ugly explorer changes (I am ready for some suggestions how do you see displaying of multiple legal identities, group them and have collapsible panel with keys to each group?)
![screenshot from 2017-08-18 15-04-05](https://user-images.githubusercontent.com/24268146/29463273-0b4ec76a-842a-11e7-8c32-b53bbf091fca.png)

mnesbit commented:
I am unclear how moving the single primary identity from NodeInfo onto ServiceHub actually tackles the multiple identity problem? The changes here touch 100+ files without giving any substantive API change.

Surely, if we are allowing nodes to have multiple context specific identities we start needing to pass in actual selection criteria at the call sites. We also need to start carrying tags, or have rules for X500 names, or make NodeInfo use an ordered List to let us select the appropriate identity according to the call context. Even if this is just a dummy "primary" string parameter it would at least allow future expansion.

@mikehearn can you give a bit more design context for this PR?

mikehearn commented:
From our chat on Friday Kasia, it sounds like you do actually agree with Matthew's comments? Is that right?

The intended goal here was to make a small set of API changes such that although we won't support multiple identities per node for now, we can add such support later without breaking API compatibility.

Now it seems like this has run aground for the following reasons:

1. Adding support for multiple identities to the data structures has ripple effects as far away as the Explorer UI.
2. We already have a kind of half-baked concept of multiple identities per node due to nodes hosting notary identities as well as their own. So attempts to take on this task end up trying to remove the special cases for notaries.
3. Changing something so fundamental involves massive changes all over the codebase. This PR already has dozens of merge conflicts.

Does that sound about right?

If it's right, then we need a new way forward here. Perhaps we need to start by changing how the notary identity special case is handled. Right now it's a "service identity" isn't it, and we'd like to remove the notion of services. Can we come up with a quick hack for now like notary traffic running on a separate port to the main legal identity of the node? And the node code largely ignores the fact that they're actually the same IP address? This should avoid us needing to use SNI for now. It doesn't scale to num_identities > num_ports of course but it should suffice for now.

Once the notion of service-ness is gone from notaries, we should be able to retire the service concept properly, clearing the path for business networks.

This would let us keep the concept of a main or default identity, which would then simplify the rest of the work. Once we add support for multiple in-node identities, we can go ahead and add the APIs that Matthew suggests to change which identity is being used. 

Does that sound approach sound better?

kasiastreich commented:
@mike, I agree with part of Matthew's comments. I can understand why you wanted data structure changes, but as Matthew pointed out it's moving a problem from one place to another. Is this an expected part of the task I’ve been assigned? I think we’d need to rethink story points allocated if so. Handling of multiple identities on the node is not trivial problem, I can see few paths where to go with that design.

As to the range of this PR. I think that we didn't realise that in every test where we do signing we use `NodeInfo.legalIdentity` for that, which is wrong. For me this usage should be moved away from `NetworkMap`, which is mainly for network purposes. We have identity service, we have Ross work on anonymised identity.

Second, we didn't specified services removal. Notary identity should be additional identity on the node (btw, it works like that even now, it's just hidden behind advertised services). We know notaries' identities form `NetworkParametes`, don't we? Then we have a problem with: what if node acts as normal peer and Oracle? Will you get it from CorDapp parameters? Oracle identity has to be different than node's "normal" one.

We should probably think if node's network TLS certificate should be mixed with node's peer identities. I will think about alternatives to using SNI.

kasiastreich commented:
@mnesbit, addressing your comment. This PR is not about multiple identity problem. It's another part of this design doc: https://r3-cev.atlassian.net/wiki/spaces/AWG/pages/127710793/Design+doc+network+map+upgrade. There is very clearly stated that implementing multiple identities on the node is not a goal of this redesign: "The only support for multi-IP/multi-identity nodes required as part of this project is slots in the data structures. Enhancing the node to support the rest of the feature is not a goal."

I could introduce choosing of one of the identites for test purposes. We already have in `ServiceHub` functions that take `publicKey` to sign.

rnicoll commented:
After a discussion with Matthew and Kasia I think we've concluded that:

1. To get this moved along, quickest path is to switch to a list of identities in a node, and use the first as the primary, to reduce impact on existing code.
2. We still need to homogenise identities a node represents (i.e. can sign contracts on behalf of), but that needs its own JIRA issue and scoping and everything.

As an addendum to 2, we definitely need some sort of node identity (for primary key in databases, for example). I'm wondering if we need the idea of an "owner identity" who is special, but not necessarily takes part in contracts, although then we have the issue of one owner having dozens of nodes and how to track each node independently. My suggestion was nodes self-issue a UUID on first run, but that's not a small change.

Anyway; @mnesbit & @kasiastreich please let me know if I've summarised acceptably.

mikehearn commented:
I think we can ignore the question of separate oracle identities for now. I don't see a problem with a flow having code like:

```
val tr = serviceHub.identityService.partyFromName("Thompson Reuters")
val rate = subFlow(com.thompsonreuters.corda.flows.GetRateFlow())
```

i.e. just hard-coding which oracle you want by the identity of the owner. Indeed, why would we want the oracle identity to be different to the normal legal identity? Given that the point of an oracle is they act as a trusted third party to sign over data, it'd be a bit odd! The legal identity is what matters most!

From Ross and Kasia's comments it sounds like using a list of identities but having the first one be special is a good way forward here, indeed. In future versions of the software we can steadily reduce the specialness of the first identity until there's no notion of a main identity anymore, or one that hardly matters. But the wire protocol and data format of the network map would remain the same, and apps can port away from the old APIs that assume "main-ness" at their own pace. Am I understanding properly?

I'm trying to feel my way through this the same as you are :)

mnesbit commented on core/src/main/kotlin/net/corda/core/flows/AbstractStateReplacementFlow.kt:
```diff
@@ -97,9 +97,8 @@ abstract class AbstractStateReplacementFlow {
         private fun collectSignatures(participants: Iterable<PublicKey>, stx: SignedTransaction): List<TransactionSignature> {
             val parties = participants.map {
-                val participantNode = serviceHub.networkMapCache.getNodeByLegalIdentityKey(it) ?:
+                serviceHub.identityService.partyFromKey(it) ?: // In identity service we record all identities we know about from network map.
```
Bit fragile leaving a comment on same line as hanging operator. Move comment up a line.

mnesbit commented on node-api/src/main/kotlin/net/corda/nodeapi/ArtemisMessagingComponent.kt:
```diff
@@ -84,7 +79,7 @@ abstract class ArtemisMessagingComponent : SingletonSerializeAsToken() {
      */
     data class ServiceAddress(val identity: PublicKey) : ArtemisAddress, MessageRecipientGroup {
-        override val queueName: String = "$SERVICES_PREFIX${identity.toBase58String()}"
+        override val queueName: String = "$PEERS_PREFIX${identity.toBase58String()}"
```
The queue name was tied to the permissions and config. Does this still work for things like RAFT notary?

mnesbit commented on node/src/main/kotlin/net/corda/node/services/messaging/ArtemisMessagingServer.kt:
```diff
@@ -302,36 +302,24 @@ class ArtemisMessagingServer(override val config: NodeConfiguration,
             }
-
-            queueName.startsWith(SERVICES_PREFIX) -> try {
-                val identity = parsePublicKeyBase58(queueName.substring(SERVICES_PREFIX.length))
```
How is the 1 to many equivalent bridging done now for CompositeKey situations?

mnesbit commented on tools/explorer/src/main/kotlin/net/corda/explorer/views/Network.kt:
```diff
@@ -88,15 +89,18 @@ class Network : CordaView() {
-                label(PartyNameFormatter.short.format(node.legalIdentity.name)) { font = Font.font(font.family, FontWeight.BOLD, 15.0) }
-                gridpane {
+                label(PartyNameFormatter.short.format(identities[0].name)) { font = Font.font(font.family, FontWeight.BOLD, 15.0) }
+                gridpane { // TODO We loose node's main identity for display :(
```
Typo lose not loose

rnicoll commented on core/src/main/kotlin/net/corda/core/messaging/CordaRPCOps.kt:
```diff
@@ -210,9 +210,9 @@ interface CordaRPCOps : RPCOps {
 
     /**
-     * Returns Node's identity, assuming this will not change while the node is running.
+     * Returns Node's NodeInfo, assuming this will not change while the node is running.
```
Needs reverting back to "identity"

rnicoll commented on core/src/main/kotlin/net/corda/core/node/NodeInfo.kt:
```diff
@@ -21,20 +21,16 @@ data class ServiceEntry(val info: ServiceInfo, val identity: PartyAndCertificate
-                    // TODO After removing of services these two fields will be merged together and made NonEmptySet.
-                    val legalIdentityAndCert: PartyAndCertificate,
-                    val legalIdentitiesAndCerts: Set<PartyAndCertificate>,
+                    val legalIdentitiesAndCerts: List<PartyAndCertificate>,
```
Could you add a note that the order is now important and that the first is the lead identity?

rnicoll commented on docs/source/example-code/src/main/java/net/corda/docs/FlowCookbookJava.java:
```diff
@@ -266,7 +266,7 @@ public Void call() throws FlowException {
             // DOCSTART 24
             DummyContract.Commands.Create commandData = new DummyContract.Commands.Create();
-            PublicKey ourPubKey = getServiceHub().getLegalIdentityKey();
+            PublicKey ourPubKey = ServiceHubKt.chooseIdentity(getServiceHub()).getOwningKey();
```
Can we avoid `chooseIdentity()` being an extension function?

shamsasari commented:
@kasiastreich as part of my changes in #1485 you will need to rename `FlowLogic.me` to `ourIdentity`. You might want to do that before rebasing
